src/services/utilServices/SgidAuthService.ts

```diff
@@ -0,0 +1,117 @@
+import SgidClient, { generatePkcePair } from "@opengovsg/sgid-client"
+import { ResultAsync, err, errAsync, okAsync } from "neverthrow"
+import { ModelStatic } from "sequelize"
+
+import { SgidLogin } from "@root/database/models/SgidLogin"
+import DatabaseError from "@root/errors/DatabaseError"
+import {
+  SgidCreateRedirectUrlError,
+  SgidFetchAccessTokenError,
+  SgidFetchUserInfoError,
+} from "@root/errors/SgidErrors"
+import logger from "@root/logger/logger"
+
+const SGID_WORK_EMAIL_SCOPE = "ogpofficerinfo.work_email"
+
+interface SgidAuthServiceProps {
+  sgidClient: SgidClient
+  sgidLoginRepository: ModelStatic<SgidLogin>
+}
+
+export default class SgidAuthService {
+  private sgidClient: SgidAuthServiceProps["sgidClient"]
+
+  private sgidLoginRepository: SgidAuthServiceProps["sgidLoginRepository"]
+
+  constructor({ sgidClient, sgidLoginRepository }: SgidAuthServiceProps) {
+    this.sgidClient = sgidClient
+    this.sgidLoginRepository = sgidLoginRepository
+  }
+
+  createSgidRedirectUrl(sessionId: string) {
+    // Generate a PKCE pair
+    const { codeChallenge, codeVerifier } = generatePkcePair()
+
+    // Generate an authorization URL
+    try {
+      const { url, nonce } = this.sgidClient.authorizationUrl({
+        state: "",
+        codeChallenge,
+        scope: ["openid", SGID_WORK_EMAIL_SCOPE],
+      })
+
+      return ResultAsync.fromPromise(
+        this.sgidLoginRepository.create({
+          state: sessionId,
```

Comment:
Technically the state that you pass in the authorization URL should match the `state` that you're storing. Perhaps rather than having the `sessionId` be the `state`, just have it as the `id`?

Reply:
After chatting with Jie Hao, we've opted to remove the db store completely - https://github.com/isomerpages/isomercms-backend/pull/860/commits/22ce8f90bb6325d20df7d4305f7688f7ff036821